\begin{abstract}
\noindent
Sequence alignment is an essential method in bioinformatics and the basis of many analyses, including phylogenetic inference, ancestral sequence reconstruction, and gene annotation.
Sequence artifacts and errors made in alignment reconstruction can impact downstream analyses leading to erroneous conclusions in comparative and functional genomic studies.
For example, abiological frameshifts and early stop codons are common artifacts found in protein coding sequences that have been annotated in reference genomes.
While such errors are eventually fixed in the reference genomes of model organisms, many genomes used by researchers contain theses artifacts, and researchers often discard large amounts of data in comparative genomic studies to prevent artifacts from impacting results.
To address this need, we present COATi, a statistical, codon-aware pairwise aligner that supports complex insertion-deletion models and can handle artifacts present in genomic data.
COATi allows users to reduce the amount of discarded data while generating more accurate sequence alignments.
\end{abstract}
