\begin{abstract}
\noindent
Sequence alignment is an essential method in bioinformatics and the basis of many analyses including phylogenetic inference, ancestral sequence reconstruction, and gene annotation. Errors made in alignment reconstruction can impact downstream analyses leading to erroneous conclusions in comparative and functional genomics.
Common artifacts in unpublished reference genomes include abiological frameshifts and early stop codons.
While for model organisms this is eventually fixed, for many species this is not the case and current curation efforts require discarding large amounts of data.
To address this need, we present COATi, a statistical codon-aware pairwise aligner that supports complex insertion-deletion models and can handle artifacts present in genomic data.
COATi will allow users to reduce the amount of discarded data while generating more accurate sequence alignments.
\end{abstract}
