\section{Methods}

Using 16000 human genes and their gorilla homologous pairs from the ENSEMBL
database \parencite{ensembl_hubbard_2002} we simulated a data set of pairwise
alignments with empirical gap patterns.
We used the data set to evaluate the accuracy of popular cutting edge aligners
Clustal$\Omega$ v1.2.4 \parencite{clustal_omega_sievers_2011},
MACSE v2.06 \parencite{ranwez_macse_2011}, MAFFT v7.407
\parencite{mafft_katoh_2002}, and PRANK v.170427 \parencite{prank_loytynoja_2014}
together with COATi fst and marginal models.

After downloading, sequences longer than 6000 nucleotides were filtered out
(2232), and 8369 alignments contained gaps identified by at least one
aligner.
We randomly introduced gap patterns extracted from all five methods into the
5399 initially ungapped sequence pairs to generate the true alignments.
Alignment accuracy was measured using the distance metric $d_{seq}$
\parencite{metrics_blackburne_whelan_2011} between simulated and inferred
alignments.
In addition, accuracy of positive and negative selection was calculated
using the $F_1$ score by estimating $k_s$ and $k_a$ statistics
\parencite{ka_ks_li_1993}.
