\section*{Results and Discussion}
Using 16000 human genes and their gorilla orthologs from the ENSEMBL
database \citembe{ensembl_hubbard_2002}, we simulated a data set of pairwise
alignments with empirical gap patterns.
We used the data set to evaluate the accuracy of COATi and a suite of popular aligners spanning multiple alignment methods:
Clustal$\Omega$ v1.2.4 \citembe{clustal_omega_sievers_2011},
MACSE v2.06 \citembe{ranwez_macse_2011}, MAFFT v7.505
\citembe{katoh2013mafft}, and PRANK v.150803 \citembe{prank_loytynoja_2014}.

After downloading, we removed 2232 sequence-pairs longer than 6000 nucleotides and then aligned the remaining pairs with all five methods.
At least one aligner added gaps to 6048 sequence pairs, and no aligner added gaps to 7719 sequence pairs.
We then randomly introduced gap patterns extracted from all five methods into the ungapped sequence pairs to generate the benchmark alignments.
The accuracy of inferred alignments compared to the benchmarks was measured using \citeauthor{metrics_blackburne_whelan_2011}'s \citeyearpar{metrics_blackburne_whelan_2011} distance metric, $d_{seq}$.
Additionally, because different alignments may be evolutionary equivalent, we scored alignments using the codon-marginal-mg model, and any alignment that had the same score as the benchmark alignment was considered equivalent to the benchmark alignment.
The accuracy of identifying positive and negative selection was calculated using the $F_1$ score by estimating $k_s$ and $k_a$ statistics
\citembe{ka_ks_li_1993} (Supplementary Methods).
The $F_1$ score evaluates the accuracy of a model by assigning equal importance to precision and recall, and ranges between 0 and 1, with a score of 1 representing a perfect result.

% Software comparison table
\begin{table}[!ht]
\centering
\input{figures/table-comp}
 \vspace{1mm}
 \footnotesize{\textsuperscript{*}PRANK produced 50 empty alignments, calculations are based on 7669 alignments.}
 \caption{COATi generates better alignments than other alignment algorithms. Results of COATi, PRANK, MAFFT, Clustal$\Omega$, and MACSE aligning 7719 empirically simulated sequence pairs. Best alignments have the lowest $d_{seq}$ (including ties), perfect alignments have the same score as the true alignment, and imperfect alignments have a different score than the true alignment when at least one method found a perfect alignment.}
 \label{table:comp}
\end{table}

COATi, using the codon-triplet-mg model, obtained better results compared to a wide variety of alignment strategies.
It was significantly more accurate (lower $d_{seq}$) at inferring the empirically simulated alignments compared to other methods; all p-values were less than $2.1 \cdot 10^{-79}$ according to the one-tailed, paired Wilcoxon signed-rank tests (Supplementary Materials Figure 1).
%
In addition, COATi produced more perfect alignments, less imperfect alignments, and more accurately inferred events of positive and negative selection (Table \ref{table:comp}).

Clustal$\Omega$ generated alignments via amino acid translations and obtained the highest average alignment error while having difficulties retrieving positive selection.
MACSE used a DNA-AA hybrid model, allowing frameshifts, and obtained similar results to MAFFT using a DNA model.
PRANK, using a codon model, had an average alignment error between MACSE/MAFFT and Clustal$\Omega$ but was unable to generate alignments for some sequence pairs.

Despite human and gorilla sequences having a relatively short evolutionary distance, COATi showed a biologically significant improvement over other methods, with an average alignment error nine-fold smaller than the next best method.
COATi is an FST-based application that can calculate the optimal alignment
between a pair of sequences in the presence of artifacts using a statistical
model.
Using COATI will allow researchers to analyze more data with higher accuracy and facilitate the study of important biological processes that shape genomic data.

Future work include extending the indel FST to combine a 3-mer gap model with a frameshift parameter and weighing each indel phase differently to reflect known selection on indel phases \citembe{zhu2022profiling}.
We also plan on comparing the marginal and triplet models to evaluate the implications of the marginalization.
