\section*{Results and Discussion}
Using 16000 human genes and their gorilla orthologs from the ENSEMBL
database \citembe{ensembl_hubbard_2002}, we simulated a data set of pairwise
alignments with empirical gap patterns.
We used the data set to evaluate the accuracy of popular aligners
Clustal$\Omega$ v1.2.4 \citembe{clustal_omega_sievers_2011},
MACSE v2.06 \citembe{ranwez_macse_2011}, MAFFT v7.505
\citembe{katoh2013mafft}, and PRANK v.150803 \citembe{prank_loytynoja_2014}
together with COATi.

After downloading, we removed 2232 sequences longer than 6000 nucleotides, identified 6048 sequence pairs that contained gaps identified by at least one aligner, and 7719 ungapped sequence pairs.
We then randomly introduced gap patterns extracted from all five methods into the ungapped sequence pairs to generate the benchmark alignments.
Alignment accuracy was measured using the distance metric $d_{seq}$
\citembe{metrics_blackburne_whelan_2011} between simulated and inferred
alignments.
In addition, accuracy of positive and negative selection was calculated
using the $F_1$ score by estimating $k_s$ and $k_a$ statistics
\citembe{ka_ks_li_1993}.
The $F_1$ score evaluates the accuracy of a model by assigning equal importance to precision and recall, and ranges between 0 and 1, with a score of 0 representing a perfect result.

% Software comparison table
\begin{table}[!ht]
\centering
\input{figures/table-comp}
 \vspace{1mm}
 \footnotesize{* PRANK produced 50 empty alignments, calculations are based on 7869 alignments.}
 \caption{COATi generates better alignments than other alignment algorithms. Results of COATi, PRANK, MAFFT, Clustal$\Omega$, and MACSE aligning 7719 empirically simulated sequence pairs. Perfect alignments have $d_{seq}=0$, best alignments have the lowest $d_{seq}$, and imperfect alignments have $d_{seq}>0$ when at least one aligner found a perfect alignment. Alignments with equal score to the true alignment are considered perfect.}
 \label{table:comp}
\end{table}

COATi, using the codon-triplet-ecm model, was significantly more accurate (lower $d_{seq}$) at inferring simulated alignments compared to other methods; all p-values were less than $2.06 \cdot 10^{-79}$ according to the one-tailed Wilcoxon signed rank test.
In addition, COATi produced more perfect alignments, less imperfect alignments, and more accurately retrieved events of positive selection (Table \ref{table:comp}).
It obtained better results compared to a wide variety of alignment strategies.
Clustal$\Omega$, performing a common approach of aligning via amino acid translations, obtained the highest average alignment error and had difficulties retrieving positive selection.
MACSE, which allows frameshifts, is also based on an amino acid model and obtained similar results to the DNA-based MAFFT.
PRANK, using a codon model, had a similar average alignment error to MACSE and MAFFT but had issues recovering the simulated alignments.

Despite human and gorilla sequences having a relatively short evolutionary distance, COATi showed a biologically significant improvement over other methods, with an average alignment error nine-fold smaller than the next best method.
COATi is an FST-based application that can calculate the optimal alignment
between a pair of sequences in the presence of artifacts using a statistical
model.
It will allow researchers to analyze more data with higher accuracy and
facilitate the study of important biological processes that shape genomic data.

Future works include extending the indel FST to combine a three-mer gap model with a frameshift parameter.
We also plan on comparing the marginal and triplet models to evaluate the implications of the marginalization.
Additionally, it would be valuable to weigh indel phases differently.
