---
title: "Supplementary Materials for \"COATi: statistical pairwise alignment of protein coding sequences\""
author: Juan J. Garcia Mesa, Ziqi Zhu, Reed A. Cartwright
output:
    pdf_document:
        fig_caption: yes
bibliography: alignpair_letter.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo = FALSE}
library(knitr)

summary_table = function(results_file, coati_model) {
    # read results_summary data
    results = read.csv(results_file)
    
    # extract dseq values for each aligner
    coati = results$dseq[which(results$aligner == coati_model)]
    clustalomega = results$dseq[which(results$aligner == "clustalo")]
    macse = results$dseq[which(results$aligner == "macse")]
    mafft = results$dseq[which(results$aligner == "mafft")]
    prank = results$dseq[which(results$aligner == "prank")]

    dseq_summary = data.frame(coati = mean(coati, na.rm = TRUE),
                      prank = mean(prank, na.rm = TRUE),
                      mafft = mean(mafft, na.rm = TRUE),
                      clustalo = mean(clustalomega, na.rm = TRUE),
                      macse = mean(macse, na.rm = TRUE),
                      row.names = "$d_{seq}$")
    
    ############################################################################
    # perfect, best, and imperfect alignments
    source("supplementary_materials/scripts/number_alignments.R")
    alns = num_alns(results_file)
    
    # fix row and column names and add to stats table
    row.names(alns)[which(row.names(alns) == coati_model)] = "coati"
    colnames(alns) = c("Perfect alignments", "Best alignments", "Imperfect alignments")
    stats_table = rbind(dseq_summary, t(alns))
    
    ############################################################################
    # selection (ka ks)
    source("supplementary_materials/scripts/kaks.R")
    
    aligners = c(coati_model, "prank", "mafft", "clustalo", "macse")
    selection = data.frame("F1.pos" = double(),
                           "F1.neg" = double())
    
    # calculate F1 score for positive and negative selection for each aligner
    for(model in aligners) {
        model_rows = results[which(results$aligner == model), ]
        pos = ps_accuracy(model_rows$ref_omega, model_rows$aln_omega)
        neg = ns_accuracy(model_rows$ref_omega, model_rows$aln_omega)
        
        selection[nrow(selection) + 1, ] = c(pos, neg)
    }
    
    # fix row and column names and add to stats table
    colnames(selection) = c("F1-score pos selection", "F1-score neg selection")
    rownames(selection) = c("coati", "prank", "mafft", "clustalo", "macse")
    stats_table = rbind(stats_table, t(selection))
    colnames(stats_table) = c(coati_model, "PRANK*", "MAFFT", "ClustalOmega", "MACSE")
    stats_table = stats_table[, c(coati_model, "MAFFT", "PRANK*", "MACSE", "ClustalOmega")]
    
    # display stats table
    stats_table
}

wilcoxon_test = function(results_file, coati_model) {
    # read results_summary data
    results = read.csv(results_file)

    # extract dseq values for each aligner
    coati = results$dseq[which(results$aligner == coati_model)]
    clustalomega = results$dseq[which(results$aligner == "clustalo")]
    macse = results$dseq[which(results$aligner == "macse")]
    mafft = results$dseq[which(results$aligner == "mafft")]
    prank = results$dseq[which(results$aligner == "prank")]

    # one-side paired wilcoxon rank test
    wtp = wilcox.test(x = coati, y = prank, paired = TRUE, alternative = "less")
    wtmf = wilcox.test(x = coati, y = mafft, paired = TRUE, alternative = "less")
    wtc = wilcox.test(x = coati, y = clustalomega, paired = TRUE, alternative = "less")
    wtmc = wilcox.test(x = coati, y = macse, paired = TRUE, alternative = "less")
    w = c(wtp$p.value, wtmf$p.value, wtc$p.value, wtmc$p.value)
    w
}
```

```{r echo = FALSE, results = 'asis'}
library(kableExtra)
library(knitr)
source("supplementary_materials/scripts/plot_dseq.R")
results = "supplementary_data/tri-mg/results/results_summary.csv"
k = kable(summary_table(results, "tri-mg"), digits = 5, caption = paste("Accuracy of COATi codon-triplet-mg, PRANK, MAFFT, ClustalOmega, and MACSE on 7719 simulated sequence pairs. Perfect alignments have the same score as the true alignment, best alignments have lowest $d_{seq}$, and imperfect alignments have a different score than the true alignment when at least one method found a perfect alignment."), escape = FALSE) %>% kable_styling(latex_options = "HOLD_position")
print(footnote(k, general = "PRANK produced 50 empty alignments, calculations are based on 7669 alignments.", footnote_as_chunk = TRUE, general_title = "*"))
#
results = "supplementary_data/tri-ecm/results/results_summary.csv"
k = kable(summary_table(results, "tri-ecm"), digits = 5, caption = paste("Accuracy of COATi codon-triplet-ecm, PRANK, MAFFT, ClustalOmega, and MACSE on 7678 simulated sequence pairs. Perfect alignments have the same score as the true alignment, best alignments have lowest $d_{seq}$, and imperfect alignments have a different score than the true alignment when at least one method found a perfect alignment."), escape = FALSE) %>% kable_styling(latex_options = "HOLD_position")
print(footnote(k, general = "PRANK produced 41 empty alignments, calculations are based on 7637 alignments.", footnote_as_chunk = TRUE, general_title = "*"))
#
results = "supplementary_data/mar-mg/results/results_summary.csv"
k = kable(summary_table(results, "mar-mg"), digits = 5, caption = paste("Accuracy of COATi codon-marginal-mg, PRANK, MAFFT, ClustalOmega, and MACSE on 7666 simulated sequence pairs. Perfect alignments have the same score as the true alignment, best alignments have lowest $d_{seq}$, and imperfect alignments have a different score than the true alignment when at least one method found a perfect alignment."), escape = FALSE) %>% kable_styling(latex_options = "HOLD_position")
print(footnote(k, general = "PRANK produced 47 empty alignments, calculations are based on 7619 alignments.", footnote_as_chunk = TRUE, general_title = "*"))
#
results = "supplementary_data/mar-ecm/results/results_summary.csv"
k = kable(summary_table(results, "mar-ecm"), digits = 5, caption = paste("Accuracy of COATi codon-marginal-ecm, PRANK, MAFFT, ClustalOmega, and MACSE on 7717 simulated sequence pairs. Perfect alignments have the same score as the true alignment, best alignments have lowest $d_{seq}$, and imperfect alignments have a different score than the true alignment when at least one method found a perfect alignment."), escape = FALSE) %>% kable_styling(latex_options = "HOLD_position")
print(footnote(k, general = "PRANK produced 40 empty alignments, calculations are based on 7677 alignments.", footnote_as_chunk = TRUE, general_title = "*"))
```

\newpage

```{r echo = FALSE}
results = "supplementary_data/tri-mg/results/results_summary.csv"
pval_tri_mg = max(wilcoxon_test(results, "tri-mg"))
results = "supplementary_data/tri-ecm/results/results_summary.csv"
pval_tri_ecm = max(wilcoxon_test(results, "tri-ecm"))
results = "supplementary_data/mar-mg/results/results_summary.csv"
pval_mar_mg = max(wilcoxon_test(results, "mar-mg"))
results = "supplementary_data/mar-ecm/results/results_summary.csv"
pval_mar_ecm = max(wilcoxon_test(results, "mar-ecm"))
```

```{r echo = FALSE, fig.width = 6, fig.height = 6, fig.align = 'center', fig.cap = paste0("\\label{fig:dseq-tri-mg} Comparison of log10-transformed $d_{seq}$ data with pseudocounts between COATi codon-triplet-mg and PRANK, MAFFT, ClustalOmega, and MACSE. COATi was significantly more accurate than other aligners; all p-values were $\\leq ", signif(pval_tri_mg, digits = 3), "$.")}
library(knitr)
source("supplementary_materials/scripts/plot_dseq.R")
suppressWarnings(plot_dseq_main("supplementary_data/tri-mg/plot_distance.csv", "tri-mg"))
```

```{r echo = FALSE, fig.width = 6, fig.height = 6, fig.align = 'center', fig.cap = paste0("\\label{fig:dseq-tri-ecm} Comparison of log10-transformed $d_{seq}$ data with pseudocounts between COATi codon-triplet-ecm and PRANK, MAFFT, ClustalOmega, and MACSE. COATi was significantly more accurate than other aligners; all p-values were $\\leq ", signif(pval_tri_ecm, digits = 3), "$.")}
library(knitr)
source("supplementary_materials/scripts/plot_dseq.R")
suppressWarnings(plot_dseq_main("supplementary_data/tri-ecm/plot_distance.csv", "tri-ecm"))
```

```{r echo = FALSE, fig.width = 6, fig.height = 6, fig.align = 'center', fig.cap = paste0("\\label{fig:dseq-mar-mg} Comparison of log10-transformed $d_{seq}$ data with pseudocounts between COATi codon-marginal-mg and PRANK, MAFFT, ClustalOmega, and MACSE. COATi was significantly more accurate than other aligners; all p-values were $\\leq ", signif(pval_mar_mg, digits = 3), "$.")}
library(knitr)
source("supplementary_materials/scripts/plot_dseq.R")
suppressWarnings(plot_dseq_main("supplementary_data/mar-mg/plot_distance.csv", "mar-mg"))
```

```{r echo = FALSE, fig.width = 6, fig.height = 6, fig.align = 'center', fig.cap = paste0("\\label{fig:dseq-mar-ecm} Comparison of log10-transformed $d_{seq}$ data with pseudocounts between COATi codon-marginal-ecm and PRANK, MAFFT, ClustalOmega, and MACSE. COATi was significantly more accurate than other aligners; all p-values were $\\leq ", signif(pval_mar_ecm, digits = 3), "$.")}
library(knitr)
source("supplementary_materials/scripts/plot_dseq.R")
suppressWarnings(plot_dseq_main("supplementary_data/mar-ecm/plot_distance.csv", "mar-ecm"))
```

\newpage

```{r echo = FALSE, fig.width = 6, fig.height = 6, fig.cap="\\label{fig:figs} Base calling error FST. Arcs from M to S generate matches; however, here they can introduce single-nucleotide errors, which can generate stop codon artifacts."}
knitr::include_graphics(path = "figures/fig-base-calling-error.pdf")
```

\newpage

# Supplementary Methods

Ks and Ka represent the number of substitutions per synonymous and non-synonymous sites. The ratio of nonsynonymous (Ka) to synonymous (Ks) nucleotide substitution rates indicates the selective pressures acting on genes. If the ratio is significantly greater than 1, it suggests positive selective pressure, meaning that nonsynonymous substitutions occur more frequently than synonymous substitutions. A ratio around 1 can indicate either neutral evolution at the protein level or a mixture of positive and negative selective pressures. If the ratio is less than 1, it indicates a pressure to maintain protein sequence, known as purifying selection.
Ks and Ka are calculated using the R package seqinr v.4.2-30 [@seqinr].

# References

<div id="refs"></div>
