---
title: "COATi: statistical pairwise alignment of protein coding sequences"
subtitle: Supplementary Materials
author: Juan J. Garcia Mesa, Ziqi Zhu, Reed A. Cartwright
mainfont: TeX Gyre Pagella
bibliography: ../alignpair_letter.bib
output:
    "bookdown::pdf_document2":
        numbered_sections: yes
        fig_caption: yes
        keep_tex: true
        latex_engine: lualatex
header-includes:
  - \let\oldthefigure\thefigure
  - \renewcommand{\thefigure}{S\oldthefigure}
  - \let\oldthetable\thetable
  - \renewcommand{\thetable}{S\oldthetable}
---

```{r setup, include = FALSE}
library(knitr)
library(tidyverse)
library(kableExtra)
library(here)
library(ggrepel)

#cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73",
#    "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")

cbbPalette <- c(
    "COATi" = "#0077BB",
    "COATi-rev" = "#33BBEE",
    "MACSE" = "#009988",
    "MAFFT" = "#EE7733",
    "PRANK" = "#CC3311",
    "ClustalΩ" = "#EE3377",
    "Benchmark" = "#000000",
    "missing" = "#BBBBBB")

set.seed(17761)

knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(dev = "cairo_pdf")

# Alignment stats
ALN_DATA <- read_csv(here("data", "raw_fasta_aligned_stats.csv.gz"))

# Preprocessed cigar strings from alignment stats
GAP_DATA <- read_csv(here("supplementary_materials", "data", "gaps.csv.gz"))
```

# Aligner Commands

We evaluated five different aligners. Below are the commands that we used to run
them. We have abbreviated the commands for clarity, stripping out unimportant
arguments. Complete workflows can be found in our coati-testing repository on
Github.

- COATi: `coati alignpair -m tri-mg ...`; We used COATi two different ways. The
  primary way used human as the reference sequence. The secondary way,
  COATi-rev, used gorilla as the reference, via a wrapper script that reversed
  the order of sequences before alignment and reversed the order back
  afterwards.
- ClustalΩ v1.2.4: `clustalo --seqtype=Protein ...`; We wrapped ClustalΩ with a
  script that translates DNA sequences into amino acid sequences (including
  stops) before alignment. Any codons that were partial or containing ambiguous
  characters were translated as "X". The script then aligned these translated
  sequences with ClustalΩ, and then created a DNA alignment that was consistent
  with the amino-acid alignment.
- MACSE v2.06: `java -jar macse.jar -prog alignSequences -seq human.fasta
  -seq_lr gorilla.fasta ...`; We wrapped MACSE with a script that created two
  temporary fasta files, one containing the human sequence and another
  containing the gorilla sequence. The human sequence was specified as the
  reliable sequence and the gorilla sequence was specified as the less-reliable
  sequence. MACSE uses "!" to mark gaps that result from frameshifts, and the
  wrapper script replaced these with "-". Additionally, MACSE sometimes produced
  columns that only contained gaps, and these columns were removed.
- PRANK v.150803: `prank -codon ...`
- MAFFT v7.520: `mafft --preservecase --globalpair --maxiterate 1000 ...`

COATi can use different alignment models for pairwise alignment. Below are the
commands that we used to run different models.

- tri-mg: `coati alignpair -m tri-mg ...`
- tri-ecm: `coati alignpair -m tri-ecm ...`
- mar-mg: `coati alignpair -m mar-mg ...`
- mar-ecm: `coati alignpair -m mar-ecm ...`

# FST Alignment Example

When using the triplet models (tri-mg and tri-ecm), COATi uses the OpenFST
library to generate best alignments by composing the input and output sequences
with the COATi FST model. While the COATi FST model is too large to display, we
can show the result of a composition.

Fig. \@ref(fig:aln-example-a) shows a graph depicting the FST that results from
composing the COATi FST with the input sequence "CTC" and the output sequence
"CTG". Every path through this FST represents one possible way to align "CTC"
and "CTG", and the sum of all weights along a path is the total weight of the
respective alignment. Here, the weights of each arc are in negative-log space.
Note that this graph has been optimized, and weight has been pushed towards the
initial state. The weight of any specific arc may not be directly mapable to a
weight described in the model.

Fig. \@ref(fig:aln-example-b) is the best alignment between "CTC" and "CTG", as
determined by the shortest path algorithm.  Figs. \@ref(fig:aln-example-a) and
\@ref(fig:aln-example-b) were produced by the OpenFST library. A bold circle
represents a starting node, and a double circle represents a termination node.

```{r aln-example-a, fig.align='center', fig.cap="The FST of all possible alignments between \"CTC\" and \"CTG\"."}
knitr::include_graphics("figures/aln-example-graph.pdf")
```

```{r aln-example-b, out.width="75%", fig.align='center', fig.cap="The best alignment of \"CTC\" and \"CTG\"."}
knitr::include_graphics("figures/aln-example-path.pdf")
```

\newpage

# Empirical Results

## Gap Patterns

### Lengths

We quantified the lengths of gaps produced by each method across the empirical
dataset of human-gorilla protein-coding-sequence pairs (Tab. \@ref
(tab:gap-table-1)). Here the gap type is either "D" for gaps introduced into
the gorilla sequence or "I" for gaps introduced in the human sequence. "COATi"
refers to the full COATi FST model with a MG substitution model
(i.e. tri-mg). "COATi-rev" refers to the same model, but using gorilla as the
reference and human as the non-reference sequence. Gap lengths of 1--6
nucleotides were binned into their respective columns. Gap lengths longer than
6 nucleotides were binned into columns 7+, 8+, and 9+ depending on whether
their lengths were 1, 2, or 3 nucleotides longer than a multiple of three.

We also quantified the lengths of gaps produced by different COATi models
(Tab. \@ref(tab:gap-table-2)).

Note that ClustalΩ gaps with lengths that are not a multiple of 3 are created by
how our wrapper script handles DNA sequences with lengths that are not multiple
of 3.

```{r gap-table-1, results = 'asis', message = FALSE}
gap_len_data <- GAP_DATA |> filter(op %in% c("I", "D")) |>
    mutate(length = if_else(length <= 6, length, 7 + ((length-1) %% 3))) |>
    arrange(length) |>
    count(method, length, op) |>
    complete(method, op, length = full_seq(length, 1), fill = list(n = 0)) |>
    pivot_wider(names_from = "length", values_from = "n")

gap_len_data <- gap_len_data |> rename("Gap Type" = "op", "Method" = "method",
    "7+" = "7", "8+" = "8", "9+" = "9")

tab <- gap_len_data |> filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method, "Gap Type")

k <- kable(tab, align = "lcrrrrrrrrr",
    caption="Number of gaps introduced by each method separated by length and type.",
    booktabs = TRUE, linesep = c("", "\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
    add_header_above(c(" " = 2L, "Gap Lengths" = 9L))
```

```{r gap-table-2, results = 'asis', message = FALSE}
tab <- gap_len_data |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method, "Gap Type") |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))

k <- kable(tab, align = "lcrrrrrrrrr",
    caption="Number of gaps introduced by different COATi models separated by length and type.",
    booktabs = TRUE, linesep = c("", "\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
    add_header_above(c(" " = 2L, "Gap Lengths" = 9L))
```

### Phases

We quantified the phases of gaps produced by different aligners and COATi
models (Tab. \@ref(tab:gap-table-3) and \@ref(tab:gap-table-4)). Phase 1 gaps
begin after the 1st position in a codon in the reference sequence, phase 2 gaps
begin after the 2nd position in a codon, and phase 3 begin after the 3rd
position in a codon (i.e. between codons). Phase 3 gaps are also known as phase
0 gaps.

Note that ClustalΩ gaps with phases of 1 and 2 are created by how our wrapper
script handles DNA sequences with lengths that are not multiple of 3.

```{r gap-table-3, results = 'asis', message = FALSE}
gap_phase_data <- GAP_DATA |> 
    mutate(mlen = if_else(op == "I", 0L, length)) |>
    mutate(pos = cumsum(mlen) - mlen, .by=c(gene, method)) |>
    mutate(phase = ((pos-1) %% 3) + 1) |>
    filter(op %in% c("I", "D")) |>
    count(method, phase) |>
    drop_na(phase) |>
    complete(method, phase = full_seq(phase, 1), fill = list(n = 0)) |>
    pivot_wider(names_from = "phase", values_from = "n")

gap_phase_data <- gap_phase_data |> rename("Method" = "method")

tab <- gap_phase_data |> filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method)

k <- kable(tab, align = "lrrr",
    caption="Number of gaps introduced by each alignment method separated by phase.",
    booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
    add_header_above(c(" " = 1L, "Gap Phases" = 3L))
```

```{r gap-table-4, results = 'asis', message = FALSE}
tab <- gap_phase_data |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))

k <- kable(tab, align = "lrrr",
    caption="Number of gaps introduced by different COATi models separated by phase.",
    booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
    add_header_above(c(" " = 1L, "Gap Phases" = 3L))
```

## Homology Patterns

We quantified the homology patterns of residues for different aligners and COATi
models (Tab. \@ref(tab:gap-table-5), Tab. \@ref(tab:gap-table-6), and
Fig. \@ref(fig:gap-figure-1)). Here we define the match, mismatch, and gap
percentages as the percent of nucleotides aligned against a match, mismatch,
and gap respectively. Note that this is different than the percent of columns
that contain a match, mismatch, or gap because match and mismatch columns are
counted twice.

```{r gap-table-5, results = 'asis', message = FALSE}
gap_count_data <- GAP_DATA |>
    count(gene, method, op, wt = length) |>
    drop_na(op) |>
    pivot_wider(id_cols=c(gene, method),
        values_from=n, names_from=op,
        values_fill = 0) |>
    rename(A = `=`)

gap_count_data <- gap_count_data |> mutate(
    anc_len = A + X + D,
    dec_len = A + X + I,
    Pa = (A + A)/(anc_len + dec_len),
    Px = (X + X)/(anc_len + dec_len),
    Pg = (D + I)/(anc_len + dec_len),
)

gap_frac_data <- gap_count_data |> 
    summarize(Matches = mean(Pa), Mismatches = mean(Px), Gaps = mean(Pg), .by = method) |>
    mutate(across(Matches:Gaps,  \(x) scales::percent(x, accuracy = 0.01))) |>
    rename(Method = method)

tab <- gap_frac_data |>
    filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method)

k <- kable(tab, align = "rccc",
    caption="Average homology percentages of alignments separated by alignment method.",
    booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position")
```

```{r gap-table-6, results = 'asis', message = FALSE}
tab <- gap_frac_data  |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))

k <- kable(tab, align = "rccc", caption="Average homology percentages of alignments separated by COATi model.", booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position")
```

```{r gap-figure-1, fig.cap="Gap fractions of COATi vs other aligners. Each panel is a scatter plot where the x coordinate is the gap fraction of a COATi alignment and y coordinate is the gap fraction of the corresponding alignment from another aligner.", message = FALSE, warning = FALSE}

# Pg

tab <- gap_count_data |>
    filter(method %in% c(c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg"))) |>
    pivot_wider(id_cols=gene, names_from=method, values_from=Pg) |>
    rename(coati = `coati-tri-mg`, `coati-rev` = `rev-coati-tri-mg`) |>
    relocate(gene, coati)

tab <- tab |> pivot_longer(cols = clustalo:`coati-rev`,
    names_to="method", values_to="other") |>
    mutate(method = fct_relevel(method, "clustalo", "macse", "mafft",
        "prank", "coati-rev")) |>
    mutate(method = fct_recode(method,
        "vs ClustalΩ" = "clustalo",
        "vs MACSE" = "macse",
        "vs MAFFT" = "mafft",
        "vs PRANK" = "prank",
        "vs COATi-rev" = "coati-rev"
    ))

pal <- cbbPalette
names(pal) <- str_c("vs ", names(pal))

gg <- ggplot(tab, aes(x = coati, y = other, color = method)) + facet_wrap(vars(method))
gg <- gg + geom_point(alpha = 0.2, show.legend = FALSE)
gg <- gg + xlab("COATi") + ylab("Other")
gg <- gg + scale_color_manual(values = pal)
gg <- gg + scale_x_continuous(trans = scales::transform_pseudo_log(0.0003, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1), limits = c(0, 1.2))
gg <- gg + scale_y_continuous(trans = scales::transform_pseudo_log(0.0003, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1))

gg <- gg + theme_grey(9) + theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_blank()
)

gg
```

\newpage

## Sequence Distances

We quantified the raw sequence distances (p-distance) inferred by different
aligners and COATi models (Fig. \@ref(fig:p-figure-1),
Tab. \@ref(tab:p-table-1), and Tab. \@ref(tab:p-table-2)). 

```{r p-figure-1, fig.cap="COATi produced shorter sequence distances than other aligners. Each panel is a scatter plot where the x coordinate is the p-distance from a COATi alignment and y coordinate is the p-distance from the corresponding alignment from another aligner.", warning = FALSE}
p_data <- ALN_DATA |>
    filter(method %in% c(c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg"))) |>
    pivot_wider(id_cols=gene, names_from=method, values_from=pdist) |>
    rename(coati = `coati-tri-mg`, `coati-rev` = `rev-coati-tri-mg`) |>
    relocate(gene, coati)

p_data1 <- p_data |> pivot_longer(cols = clustalo:`coati-rev`,
    names_to="method", values_to="other") |>
    mutate(method = fct_relevel(method, "clustalo", "macse", "mafft",
        "prank", "coati-rev")) |>
    mutate(method = fct_recode(method,
        "vs ClustalΩ" = "clustalo",
        "vs MACSE" = "macse",
        "vs MAFFT" = "mafft",
        "vs PRANK" = "prank",
        "vs COATi-rev" = "coati-rev"
    ))

pal <- cbbPalette
names(pal) <- str_c("vs ", names(pal))

gg <- ggplot(p_data1, aes(x = coati, y = other, color = method)) + facet_wrap(vars(method))
gg <- gg + geom_point(alpha = 0.2, show.legend = FALSE)
gg <- gg + xlab("COATi") + ylab("Other")
gg <- gg + scale_color_manual(values = pal)
gg <- gg + scale_x_continuous(trans = scales::transform_pseudo_log(0.0004, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1))
gg <- gg + scale_y_continuous(trans = scales::transform_pseudo_log(0.0004, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1))

gg <- gg + theme_grey(9) + theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_blank()
)

gg
```

```{r p-table-1, results = 'asis', message = FALSE}
tab <- ALN_DATA |> summarize(m = mean(pdist, na.rm = TRUE), .by = "method") |>
    rename(Method = method)

tab_data1 <- tab |>
    filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method) |>
    pivot_wider(names_from = "Method", values_from = "m")

k <- kable(tab_data1, digits = 4, align = "ccccc",
  caption="Average p-distance of alignments for each method.",
  booktabs = TRUE, linesep = c("\\addlinespace"),
  format.args = list(nsmall = 4))

k |> kable_styling(latex_options = "HOLD_position")
```

```{r p-table-2, results = 'asis', message = FALSE}
tab_data2 <- tab |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))  |>
    pivot_wider(names_from = "Model", values_from = "m")

k <- kable(tab_data2, digits = 4, align = "ccccc",
    caption="Average p-distance of alignments for each COATi model.",
    booktabs = TRUE, linesep = c("\\addlinespace"),
    format.args = list(nsmall = 4))

k |> kable_styling(latex_options = "HOLD_position")
```

\newpage

## Evolutionary Distances

We quantified the evolutionary distances inferred by different aligners and
COATi models (Fig. \@ref(fig:k2p-figure-1), Tab. \@ref(tab:k2p-table-1), and
Tab. \@ref(tab:k2p-table-2)). 

```{r k2p-figure-1, fig.cap="COATi produced shorter evolutionary distances than other aligners. Each panel is a scatter plot where the x coordinate is the K2P distance from a COATi alignment and y coordinate is the K2P distance from the corresponding alignment from another aligner.", warning = FALSE}
k2p_data <- ALN_DATA |>
    filter(method %in% c(c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg"))) |>
    pivot_wider(id_cols=gene, names_from=method, values_from=k2p) |>
    rename(coati = `coati-tri-mg`, `coati-rev` = `rev-coati-tri-mg`) |>
    relocate(gene, coati)

k2p_data1 <- k2p_data |> pivot_longer(cols = clustalo:`coati-rev`,
    names_to="method", values_to="other") |>
    mutate(method = fct_relevel(method, "clustalo", "macse", "mafft",
        "prank", "coati-rev")) |>
    mutate(method = fct_recode(method,
        "vs ClustalΩ" = "clustalo",
        "vs MACSE" = "macse",
        "vs MAFFT" = "mafft",
        "vs PRANK" = "prank",
        "vs COATi-rev" = "coati-rev"
    ))

pal <- cbbPalette
names(pal) <- str_c("vs ", names(pal))

gg <- ggplot(k2p_data1, aes(x = coati, y = other, color = method)) + facet_wrap(vars(method))
gg <- gg + geom_point(alpha = 0.2, show.legend = FALSE)
gg <- gg + xlab("COATi") + ylab("Other")
gg <- gg + scale_color_manual(values = pal)
gg <- gg + scale_x_continuous(trans = scales::transform_pseudo_log(0.0004, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1))
gg <- gg + scale_y_continuous(trans = scales::transform_pseudo_log(0.0004, base=10),
    breaks=c(0, 0.001, 0.01, 0.1, 1))

gg <- gg + theme_grey(9) + theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_blank()
)

gg
```

```{r k2p-table-1, results = 'asis', message = FALSE}
tab <- ALN_DATA |> summarize(m = mean(k2p, na.rm = TRUE), .by = "method") |>
    rename(Method = method)

tab_data1 <- tab |>
    filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method) |>
    pivot_wider(names_from = "Method", values_from = "m")

k <- kable(tab_data1, digits = 4, align = "ccccc",
  caption="Average K2P-distance of alignments for each method.",
  booktabs = TRUE, linesep = c("\\addlinespace"),
  format.args = list(nsmall = 4))

k |> kable_styling(latex_options = "HOLD_position")
```

```{r k2p-table-2, results = 'asis', message = FALSE}
tab_data2 <- tab |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))  |>
    pivot_wider(names_from = "Model", values_from = "m")

k <- kable(tab_data2, digits = 4, align = "ccccc",
    caption="Average K2P-distance of alignments for each COATi model.",
    booktabs = TRUE, linesep = c("\\addlinespace"),
    format.args = list(nsmall = 4))

k |> kable_styling(latex_options = "HOLD_position")
```

# Benchmark Results

## Alignment Distances

For each sequence pair in the benchmark, we calculated the alignment distance
(d~seq~) between the benchmark alignment and the alignments generated by COATi,
ClustalΩ, MACSE, MAFFT, and PRANK. We also calculated distances between all
pairs of aligners. Our benchmark contained gap patterns extracted from
alignments generated by different aligners. Figure
\@ref(fig:bench-pcoa-split) contains the results of a metric multidimensional
scaling (principle coordinate analysis; PCoA) of the matrix of average
distances between aligners, separated by what type of gap patterns was used in
the benchmark alignment.

Figure \@ref(fig:bench-pcoa-all) contains a principle coordinate analysis of
each aligner, including different COATi models, across the entire benchmark
dataset. COATi's different models produced similar alignments and cluster
together along with the benchmarks.

```{r bench-pcoa-split, message = FALSE, fig.cap="COATi produced accurate alignments regardless of whether the underlying gap pattern was extracted from an alignment generated by another program. Each panel is a metric multidimensional scaling of the average distances between aligners."}
tab_dist <- read_csv(here("supplementary_materials", "data", "benchmarks-dseq-split.csv"))

methods <- unique(c(tab_dist$method1, tab_dist$method2))

tab_xy <- tab_dist |> group_by(origin) |> 
    group_modify(\(df, k) {
        m <- matrix(0, length(methods), length(methods))
        m[lower.tri(m)] <- df$dseq
        colnames(m) <- methods
        rownames(m) <- methods
        d <- as.dist(m)
        v <- cmdscale(d, 2)
        colnames(v) <- c("x", "y")
        as_tibble(v, rownames = "method")
    }) |> ungroup()

method_recode <- c(
        "Benchmark" = "benchmark",
        "COATi" = "coati-tri-mg",
        "ClustalΩ" = "clustalo",
        "MACSE" = "macse",
        "MAFFT" = "mafft",
        "PRANK" = "prank"
)

origin_recode <- c(
        "COATi Gaps" = "coati-tri-mg",
        "ClustalΩ Gaps" = "clustalo",
        "MACSE Gaps" = "macse",
        "MAFFT Gaps" = "mafft",
        "PRANK Gaps" = "prank"
)

tab <- tab_xy |> mutate(
    method = fct_recode(method, !!!method_recode),
    origin = fct_recode(origin, !!!origin_recode) )

pal <- cbbPalette

gg1 <- ggplot(tab, aes(x = x, y = y, label = method, color =  method)) +
    geom_point(show.legend = FALSE) +
    geom_text_repel(show.legend = FALSE, seed = 3)

gg1 <- gg1 + facet_wrap(vars(origin))

gg1 <- gg1 + xlab("Coordinate 1") + ylab("Coordinate 2")

gg1 <- gg1 + scale_color_manual(values = pal)

gg1 <- gg1 + coord_fixed()

gg1 <- gg1 + theme_grey(9)

gg1

```

```{r bench-pcoa-all, message = FALSE, fig.cap="Different COATi models produce accurate alignments that are also similar to one another."}
tab_dist <- read_csv(here("supplementary_materials", "data", "benchmarks-dseq.csv"))

methods <- unique(c(tab_dist$method1, tab_dist$method2))

m <- matrix(0, length(methods), length(methods))
m[lower.tri(m)] <- tab_dist $dseq
colnames(m) <- methods
rownames(m) <- methods
d <- as.dist(m)
v <- cmdscale(d, 2)
colnames(v) <- c("x", "y")
tab_xy <- as_tibble(v, rownames = "method")

method_recode <- c(
    "Benchmark" = "benchmark",
    "COATi-tri-mg" = "coati-tri-mg",
    "COATi-tri-ecm" = "coati-tri-ecm",
    "COATi-mar-mg" = "coati-mar-mg",
    "COATi-mar-ecm" = "coati-mar-ecm",
    "COATi-dna" = "coati-dna",
    "COATi-rev-tri-mg" = "rev-coati-tri-mg",
    "ClustalΩ" = "clustalo",
    "MACSE" = "macse",
    "MAFFT" = "mafft",
    "PRANK" = "prank"
)

tab <- tab_xy |> mutate(
    method = fct_recode(method, !!!method_recode))

pal <- c(
    "COATi-tri-mg" = "#0077BB",
    "COATi-tri-ecm" = "#0077BB",
    "COATi-mar-mg" = "#0077BB",
    "COATi-mar-ecm" = "#0077BB",
    "COATi-dna" = "#0077BB",
    "COATi-rev-tri-mg" = "#33BBEE",
    "MACSE" = "#009988",
    "MAFFT" = "#EE7733",
    "PRANK" = "#CC3311",
    "ClustalΩ" = "#EE3377",
    "Benchmark" = "#000000",
    "missing" = "#BBBBBB")

gg <- ggplot(tab, aes(x = x, y = y, label = method, color =  method)) +
    geom_point(show.legend = FALSE) +
    geom_text_repel(show.legend = FALSE, seed = 1, max.overlaps = 20, force = 2)

gg <- gg + xlab("Coordinate 1") + ylab("Coordinate 2")
gg <- gg + scale_color_manual(values = pal)
gg <- gg + coord_fixed()
gg <- gg + theme_grey(9)

gg
```

<!-- # References -->

<!-- <div id="refs"></div> -->
