---
title: "COATi: statistical pairwise alignment of protein coding sequences"
subtitle: Supplementary Materials
author: Juan J. Garcia Mesa, Ziqi Zhu, Reed A. Cartwright
mainfont: TeX Gyre Pagella
bibliography: ../alignpair_letter.bib
output:
    "bookdown::pdf_document2":
        numbered_sections: yes
        fig_caption: yes
        keep_tex: true
        latex_engine: lualatex
header-includes:
  - \let\oldthefigure\thefigure
  - \renewcommand{\thefigure}{S\oldthefigure}
  - \let\oldthetable\thetable
  - \renewcommand{\thetable}{S\oldthetable}
---

```{r setup, include = FALSE}
library(knitr)
library(tidyverse)
library(kableExtra)
library(here)

knitr::opts_chunk$set(echo = FALSE)
```

```{r}

summary_table = function(results_file, coati_model) {
    # read results_summary data
    results = read.csv(results_file)
    
    # extract dseq values for each aligner
    coati = results |> filter(aligner == coati_model) |> select(dseq)
    clustal = results |> filter(aligner == "clustalo") |> select(dseq)
    macse = results |> filter(aligner == "macse") |> select(dseq)
    mafft = results |> filter(aligner == "mafft") |> select(dseq)
    prank = results |> filter(aligner == "prank") |> select(dseq)
    
    stats_table = data.frame(dseq = c(mean(coati$dseq, na.rm = TRUE),
                          mean(mafft$dseq, na.rm = TRUE),
                          mean(prank$dseq, na.rm = TRUE),
                          mean(macse$dseq, na.rm = TRUE),
                          mean(clustal$dseq, na.rm = TRUE)), row.names = c(coati_model, "mafft", "prank", "macse", "clustalo"))
    
    ###########################################################################
    # perfect, best, and imperfect alignments
    source("scripts/number_alignments.R")
    alns = num_alns(results_file)
    
    # fix row and column names and add to stats table
    colnames(alns) = c("Perfect alns", "Best alns", "Imperfect alns")
    alns = alns[c(coati_model, "mafft", "prank", "macse", "clustalo"), ]
    stats_table = cbind(stats_table, alns)
    
    ############################################################################
    # selection (ka ks)
    source("scripts/kaks.R")
    
    aligners = c(coati_model, "mafft", "prank", "macse", "clustalo")
    selection = data.frame("F1.pos" = double(),
                           "F1.neg" = double())
    
    # calculate F1 score for positive and negative selection for each aligner
    for(model in aligners) {
        model_rows = results[which(results$aligner == model), ]
        pos = ps_accuracy(model_rows$ref_omega, model_rows$aln_omega)
        neg = ns_accuracy(model_rows$ref_omega, model_rows$aln_omega)
        
        selection[nrow(selection) + 1, ] = c(pos, neg)
    }
    
    # fix row and column names and add to stats table
    colnames(selection) = c("F1 pos selection", "F1 neg selection")
    rownames(selection) = c("coati", "mafft", "prank", "macse", "clustalo")
    stats_table = cbind(stats_table, selection)
    if(coati_model == "tri-mg") {
        coati_model = "Triplet-MG94"
    } else if(coati_model == "tri-ecm") {
        coati_model = "Triplet-ECM"
    } else if(coati_model == "mar-mg") {
        coati_model = "Marginal-MG94"
    } else if(coati_model == "mar-ecm") {
        coati_model = "Marginal-ECM"
    }
    rownames(stats_table) = c(coati_model, "MAFFT", "PRANK*", "MACSE", "ClustalOmega")
    
    # display stats table
    stats_table
}

wilcoxon_test = function(results_file, coati_model) {
    # read results_summary data
    results = read.csv(results_file)

    # extract dseq values for each aligner
    coati = results$dseq[which(results$aligner == coati_model)]
    clustalomega = results$dseq[which(results$aligner == "clustalo")]
    macse = results$dseq[which(results$aligner == "macse")]
    mafft = results$dseq[which(results$aligner == "mafft")]
    prank = results$dseq[which(results$aligner == "prank")]

    # one-side paired wilcoxon rank test
    wtp = wilcox.test(x = coati, y = prank, paired = TRUE, alternative = "less")
    wtmf = wilcox.test(x = coati, y = mafft, paired = TRUE, alternative = "less")
    wtc = wilcox.test(x = coati, y = clustalomega, paired = TRUE, alternative = "less")
    wtmc = wilcox.test(x = coati, y = macse, paired = TRUE, alternative = "less")
    w = c(wtp$p.value, wtmf$p.value, wtc$p.value, wtmc$p.value)
    w
}
```

# Aligner Commands

We evaluated five different aligners. Below are the commands that we used to run them. We have abbreviated the commands for clarity, stripping out unimportant arguments. Complete workflows can be found in our coati-testing repository on Github.

- COATi: `coati alignpair -m tri-mg ...`; We used COATi two different ways. The primary way used human as the reference sequence. The secondary way, COATi-rev, used gorilla as the reference, via a wrapper script that reversed the order of sequences before alignment and reversed the order back afterwards.
- ClustalΩ v1.2.4: `clustalo --seqtype=Protein ...`; We wrapped ClustalΩ with a script that translates DNA sequences into amino acid sequences (including stops) before alignment. Any codons that were partial or containing ambiguous characters were translated as "X". The script then aligned these translated sequences with ClustalΩ, and then created a DNA alignment that was consistent with the amino-acid alignment.
- MACSE v2.06: `java -jar macse.jar -prog alignSequences -seq human.fasta -seq_lr gorilla.fasta ...`; We wrapped MACSE with a script that created two temporary fasta files, one containing the human sequence and another containing the gorilla sequence. The human sequence was specified as the reliable sequence and the gorilla sequence was specified as the less-reliable sequence. MACSE uses "!" to mark gaps that result from frameshifts, and the wrapper script replaced these with "-". Additionally, MACSE sometimes produced columns that only contained gaps, and these columns were removed.
- PRANK v.150803: `prank -codon ...`
- MAFFT v7.520: `mafft --preservecase --globalpair --maxiterate 1000 ...`

COATi can use different alignment models for pairwise alignment. Below are the commands that we used to run different models.

- tri-mg: `coati alignpair -m tri-mg ...`
- tri-ecm: `coati alignpair -m tri-ecm ...`
- mar-mg: `coati alignpair -m mar-mg ...`
- mar-ecm: `coati alignpair -m mar-ecm ...`

# FST Alignment Example

When using the triplet models (tri-mg and tri-ecm), COATi uses the OpenFST library to generate best alignments by composing the input and output sequences with the COATi FST model. While the COATi FST model is too large to display, we can show the result of a composition.

Fig. \@ref(fig:aln-example-a) shows a graph depicting the FST that results from composing the COATi FST with the input sequence "CTC" and the output sequence "CTG". Every path through this FST represents one possible way to align "CTC" and "CTG", and the sum of all weights along a path is the total weight of the respective alignment. Here, the weights of each arc are in negative-log space. Note that this graph has been optimized, and weight has been pushed towards the initial state. The weight of any specific arc may not be directly mapable to a weight described in the model.

Fig. \@ref(fig:aln-example-b) is the best alignment between "CTC" and "CTG", as determined by the shortest path algorithm.  Figs. \@ref(fig:aln-example-a) and \@ref(fig:aln-example-b) were produced by the OpenFST library. A bold circle represents a starting node, and a double circle represents a termination node.

```{r aln-example-a, fig.align='center', fig.cap="The FST of all possible alignments between `CTC' and `CTG'."}
knitr::include_graphics("figures/aln-example-graph.pdf")
```

```{r aln-example-b, out.width="75%", fig.align='center', fig.cap="The best alignment of `CTC' and `CTG'."}
knitr::include_graphics("figures/aln-example-path.pdf")
```

\newpage

# Gap Patterns

## Lengths

We quantified the lengths of gaps produced by each method across the empirical dataset of human-gorilla protein-coding-sequence pairs (Tab. \@ref(tab:gap-table-1)). Here the gap type is either "D" for gaps introduced into the gorilla sequence or "I" for gaps introduced in the human sequence. "COATi" refers to the full COATi FST model with a MG substitution model (i.e. tri-mg). "COATi-rev" refers to the same model, but using gorilla as the reference and human as the non-reference sequence. Gap lengths of 1--6 nucleotides were binned into their respective columns. Gap lengths longer than 6 nucleotides were binned into columns 7+, 8+, and 9+ depending on whether their lengths were 1, 2, or 3 nucleotides longer than a multiple of three.

We also quantified the lengths of gaps produced by different COATi models (Tab. \@ref(tab:gap-table-2)).

Note that ClustalΩ gaps with lengths that are not a multiple of 3 are created by how our wrapper script handles DNA sequences with lengths that are not multiple of 3.

```{r gap-table-1, results = 'asis', message = FALSE}
gap_data <- read_csv(here("supplementary_materials/data/gap-lengths.csv"))

gap_data <- gap_data |> rename("Gap Type" = "values", "Method" = "method",
  "7+" = "7", "8+" = "8", "9+" = "9")

gap_data1 <- gap_data |> filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method, "Gap Type")

k <- kable(gap_data1, align = "lcrrrrrrrrr", caption="Number of gaps introduced by each method separated by length and type.", booktabs = TRUE, linesep = c("", "\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
  add_header_above(c(" " = 2L, "Gap Lengths" = 9L))
```

```{r gap-table-2, results = 'asis', message = FALSE}
gap_data2 <- gap_data |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method, "Gap Type") |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))

k <- kable(gap_data2, align = "lcrrrrrrrrr", caption="Number of gaps introduced by different COATi models separated by length and type.", booktabs = TRUE, linesep = c("", "\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
  add_header_above(c(" " = 2L, "Gap Lengths" = 9L))
```

## Phases

We quantified the phases of gaps produced by different aligners and COATi models (Tab. \@ref(tab:gap-table-3) and \@ref(tab:gap-table-4)). Phase 1 gaps begin after the 1st position in a codon in the reference sequence, phase 2 gaps begin after the 2nd position in a codon, and phase 3 begin after the 3rd position in a codon (i.e. between codons). Phase 3 gaps are also known as phase 0 gaps.

Note that ClustalΩ gaps with phases of 1 and 2 are created by how our wrapper script handles DNA sequences with lengths that are not multiple of 3.

```{r gap-table-3, results = 'asis', message = FALSE}
gap_data <- read_csv(here("supplementary_materials/data/gap-phases.csv"))

gap_data <- gap_data |> rename("Method" = "method")

gap_data1 <- gap_data |> filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method)

k <- kable(gap_data1, align = "lrrr", caption="Number of gaps introduced by each alignment method separated by phase.", booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
  add_header_above(c(" " = 1L, "Gap Phases" = 3L))
```

```{r gap-table-4, results = 'asis', message = FALSE}
gap_data2 <- gap_data |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))

k <- kable(gap_data2, align = "lrrr", caption="Number of gaps introduced by different COATi models separated by phase.", booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position") |>
  add_header_above(c(" " = 1L, "Gap Phases" = 3L))
```

## Gapiness

We quantified the gapiness of different aligners and COATi models (Tab. \@ref(tab:gap-table-5) and \@ref(tab:gap-table-6)). Here we define gapiness as the tendency of an aligner to introduce gaps into an alignment, and we measure it by calculating the percentage of nucleotides aligned with a gap. Note that this is different than the percent of columns that contain a gap.

```{r gap-table-5, results = 'asis', message = FALSE}
gap_data <- read_csv(here("supplementary_materials/data/gapiness.csv"))

gap_data <- gap_data |> rename("Method" = "method") |>
    mutate("Gapiness" = sprintf("%0.2f%%",round(p * 100, 2))) |>
    select(-p)

gap_data1 <- gap_data |>
    filter(Method %in% c("clustalo", "coati-tri-mg",
    "macse", "mafft", "prank", "rev-coati-tri-mg")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg", "rev-coati-tri-mg"))) |>
    mutate(Method = case_match(Method,
        "clustalo" ~ "ClustalΩ",
        "coati-tri-mg" ~ "COATi",
        "macse" ~ "MACSE",
        "mafft" ~ "MAFFT",
        "prank" ~ "PRANK",
        "rev-coati-tri-mg" ~ "COATi-rev"
        )) |>
    arrange(Method) |>
    pivot_wider(names_from = "Method", values_from = "Gapiness")

k <- kable(gap_data1, align = "ccccc", caption="Average gapiness of alignments separated by alignment method.", booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position")
```

```{r gap-table-6, results = 'asis', message = FALSE}
gap_data2 <- gap_data |> filter(Method %in% c("coati-tri-mg",
    "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna")) |>
    mutate(Method = fct_relevel(Method, c("coati-tri-mg",
        "coati-tri-ecm", "coati-mar-mg", "coati-mar-ecm", "coati-dna"))) |>
    arrange(Method) |>
    rename(Model = Method) |>
    mutate(Model = fct_recode(Model,
        "TRI-MG" = "coati-tri-mg",
        "TRI-ECM" = "coati-tri-ecm",
        "MAR-MG" = "coati-mar-mg",
        "MAR-ECM" = "coati-mar-ecm",
        "DNA" = "coati-dna"
    ))  |>
    pivot_wider(names_from = "Model", values_from = "Gapiness")

k <- kable(gap_data2, align = "ccccc", caption="Average gapiness of alignments separated by COATi model.", booktabs = TRUE, linesep = c("\\addlinespace"))

k |> kable_styling(latex_options = "HOLD_position")
```

<!-- # References -->

<!-- <div id="refs"></div> -->
